Periodically refresh known/preferred peer IPs

marta-lokhova opened the pull request:
Resolves #1908.
note that as part of this we now avoid resolving preferred peer IPs twice in `storeConfigPeers`

vogel commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -257,9 +261,43 @@ OverlayManagerImpl::start()
 }
 
+PeerBareAddress
+OverlayManagerImpl::getLatestPeerAddress(std::string const& peerStr,
```
That name is misleading, even if this method return the lastest address.
Problem is that is also updates peer cache, updates peer table and possibly updates mConfigurationPreferredPeers, so it is much more than getting that address.

My proposition is to make it responsible for less things.

1. isPreferred parameter could be removed, and updating mConfigurationPreferredPeers could be done in maybeRefreshPeerIPs
2. name could be changed to something like updatePeerAddressCache
3. call to addAddressToExistingPeer could also be moved to maybeRefreshPeerIPs, and condition if (storedAddress != address) should be added to PeerManager::addAddressToExistingPeer

marta-lokhova replied:
What do you think about instead renaming this function to something like `resolveIPAndMaybeUpdateCaches`? The reason I don't like splitting it is because all cache/db updates are consolidated in one place, so it's less overhead to track these. Also, `storedAddress != address` condition prevents from redundant cache inserts.

vogel replied:
I think that with method's name so long it may be better to extract small utility class just for cache handling (if the cache will be still there, after resolving Nicolas comments).

vogel commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -391,6 +426,34 @@ OverlayManagerImpl::connectTo(std::vector<PeerBareAddress> const& peers,
+    }
+    catch (std::runtime_error& e)
+    {
+        CLOG(ERROR, "Overlay")
```
This fails after just the first error and does not let other peers get resolved even if the error only occurs on first one. Moving this try/catch inside getLatestPeerAddress would fix that issue.

vogel commented:
Idea with setting peer type to OUTBOUND in storeConfigPeers is a good one.

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
         {
-            CLOG(ERROR, "Overlay")
-                << "Unable to add preferred peer '" << s << "'";
+            CLOG(ERROR, "Overlay") << "Unable to resolve peer '" << peer << "'";
```
you should probably include the error message from `e` in the error

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -253,7 +256,11 @@ OverlayManagerImpl::start()
             [this]() {
-                storeConfigPeers();
+                storeConfigPeers(true);
+                mPeerIPTimer.expires_from_now(PEER_IP_RESOLVE_DELAY);
```
I wonder if we should just call this code from `tick` instead of using a dedicated timer? Less things to keep track of

MonsieurNicolas replied:
Alternatively, you could just call
```
storeConfigPeers(true);
storeConfigPeers(false);
```
to avoid duplicating the scheduling.

marta-lokhova replied:
I ended up going with your latter suggestion. With the former I think we'd end up resolving peers too often (although we can check timestamps in tick to resolve only after some time, but that would use a timer anyway)

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
 void
-OverlayManagerImpl::storePeerList(std::vector<std::string> const& list,
-                                  bool setPreferred)
+OverlayManagerImpl::storePeerList(std::vector<PeerBareAddress> addresses,
```
pass `std::vector<PeerBareAddress> const& addresses` instead of by value

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
                            : PeerManager::TypeUpdate::UPDATE_TO_OUTBOUND;
-
-    for (auto const& peerStr : list)
+    for (auto const& peer : addresses)
```
you should probably clear `mConfigurationPreferredPeers` before the loop if we're setting preferred peers as it's supposed to represent the peers from config

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
 void
-OverlayManagerImpl::storeConfigPeers()
+OverlayManagerImpl::storeKnownAndPreferred(
+    std::vector<PeerBareAddress> known, std::vector<PeerBareAddress> preferred,
```
pass by const reference -> don't need to `std::move`

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
+    else
+    {
+        mApp.postOnBackgroundThread(
+            [this, startup]() {
```
don't need `startup` as it's `false`

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
+    {
+        mApp.postOnBackgroundThread(
+            [this, startup]() {
+                auto known = resolvePeers(this->mApp.getConfig().KNOWN_PEERS);
```
as we're on a different thread...
1. how do we guarantee that `this` is valid? 
2. how do we ensure that things don't blow up on shutdown? In particular, is it safe to have this run for a long time while the main thread is tearing down overlay?

MonsieurNicolas replied:
looks to me that a pattern that uses `std::future` would be cleaner and easier to reason about, you could `peek` into this future from the `tick` cb

marta-lokhova replied:
Agreed that `std::future` seems to be safer in this case. 
To address your questions:
1. From what I see, when application is destructing, it first joins all threads (while blocking), before destroying its member variables. This means that the background thread will finish its work while the app (hence `mOverlayManager`, and hence captured `this`) is still alive. 
2. I think the right course of action here would be to stop any activity as soon as OverlayManager starts shutting down, i.e. stop ticking and resolving any peers.

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
+                auto preferred =
+                    resolvePeers(this->mApp.getConfig().PREFERRED_PEERS);
+
+                this->mApp.postOnMainThread(
```
related to shutdown: can we actually post on main thread if overlay is shutting down?

marta-lokhova replied:
(Marking this as resolved as this is not an issue anymore with `std::future` approach)

MonsieurNicolas commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
+}
+
+std::vector<PeerBareAddress>
+OverlayManagerImpl::resolvePeers(std::vector<std::string> peers)
```
pass by const reference

MonsieurNicolas commented on src/overlay/OverlayManagerTests.cpp:
```diff
@@ -118,24 +118,89 @@ class OverlayManagerTests
 
     void
-    test_addPeerList()
+    test_addPeerList(bool async = false)
```
as you're touching this, can you rename the `test_*` functions to use something closer to our naming convention?

vogel commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
+};
+
+void
+OverlayManagerImpl::storeConfigPeers(bool startup)
```
This function has two branches that have nothing in common, so it will be better to just have two functions here (storeConfigPeersSync and storeConfigPeersAsync?).

marta-lokhova replied:
Sure, updated to `storeConfigPeers` and `triggerPeerResolution`.

vogel commented:
When we look at the whole picture, there are some members and methods tightly coupled with each other:

- storePeerList
- storeKnownAndPreferred
- storeConfigPeers
- resolvePeers
- mConfigurationPreferredPeers

Exctracting them to separate utility class would make OverlayManagerImpl smaller and more mainteinable, also creating new entity with clear name and purpose (ConfigPeerResolver?).

vogel commented on src/overlay/OverlayManagerImpl.cpp:
```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
-            CLOG(ERROR, "Overlay") << "Unable to add peer '" << peerStr << "'";
+            // If address is present in the DB, `update` will ensure
+            // type is correctly updated. Otherwise, a new entry is created.
+            // Note that this won't downgrade promoted peers back to outbound.
```
typo, should be:

Note that this won't downgrade *preferred* peers back to outbound.

marta-lokhova commented:
Updated this, main change being the use of `std::future` instead of posting back to main thread.

@vogel wanted to follow up on your comment -- some code got moved around, and I removed `storeKnownAndPreferred`. With the updated code, do you think it's still worth separating the functions/members you mentioned? I also realize that overlay tests depend a lot on the privates of OverlayManager, as OverlayManagerTests is a friend of OverlayManager :(

vogel commented:
In my opinion it is almost always worth to separate members - it allows for smaller responsibilities, better testability and readability and also enforces encapsulation. I also think that the need of OverlayManagerTests is a result of OverlayManager being too big in general (how to unit test a class that has 30 methods and that depends on Application class that is king of [god object](https://en.wikipedia.org/wiki/God_object) itself?).

I would vote for separation, but I'm not sure this idea would get majority of votes.

marta-lokhova commented:
The problem is that I think OverlayManager itself is kind of tightly coupled with the functionality you mentioned (like it decides whether resolution should be sync or async, it relies on mConfigurationPreferredPeers, prevents resolving when shutting down, etc)

MonsieurNicolas commented:
I think the cleanup idea is not in scope for this PR. I am r+ this as it is

MonsieurNicolas commented:
r+ 54affb902158338874f2ee1e90bf702b3671dac5
